File: conversions/array-unique/elide-array-unique.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-array-unique
#
# This demonstrates calling the TypeScript unique implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One unique library shared across TypeScript and Ruby
# - Consistent deduplication across microservices
# - No need for custom Ruby uniq logic
# - Perfect for Rails data processing

puts "=== Ruby Calling TypeScript Array Unique ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# unique_module = Elide.require('./elide-array-unique.ts')

# Example 1: Rails Model Deduplication
# class TagService
#   def get_unique_tags(articles)
#     all_tags = articles.flat_map(&:tags)
#     unique_module.default(all_tags)
#   end
# end

# Example 2: Data Processing
# class DataProcessor
#   def remove_duplicates(data)
#     unique_module.default(data)
#   end
# end

puts "Real-world use case:"
puts "- Ruby Rails app deduplicates data"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent unique behavior"
puts "- No Array.uniq needed"
puts ""

puts "Example: Data Deduplication System"
puts "┌─────────────────────────────────────────────┐"
puts "│   Elide Array Unique (TypeScript)         │"
puts "│   elide-array-unique.ts                    │"
puts "└─────────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same deduplication logic everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different unique implementations = order inconsistencies"
puts "After: One Elide implementation = unified deduplication"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/unique.rb"
puts "  UNIQUE = Elide.require('./elide-array-unique.ts')"
puts "  "
puts "  # app/services/tag_service.rb"
puts "  def unique_tags(tags)"
puts "    UNIQUE.default(tags)"
puts "  end"


# ============================================================
# Extended Usage Examples
# ============================================================

# Example: Rails Service Object
# class ProcessingService
#   def initialize
#     @module = Elide.require('./elide-array-unique.ts')
#   end
#
#   def process(data)
#     @module.default(data)
#   end
#
#   def process_batch(items)
#     items.map { |item| process(item) }
#   end
# end

# Example: Rails Controller Integration
# class DataController < ApplicationController
#   before_action :load_module
#
#   def process
#     data = params.require(:data)
#     result = @module.default(data)
#     render json: { result: result }
#   rescue => e
#     render json: { error: e.message }, status: :unprocessable_entity
#   end
#
#   private
#
#   def load_module
#     @module = Elide.require('./elide-array-unique.ts')
#   end
# end

# Example: Background Job
# class DataProcessorJob < ApplicationJob
#   queue_as :default
#
#   def perform(data_id)
#     data = Data.find(data_id)
#     module = Elide.require('./elide-array-unique.ts')
#     result = module.default(data.content)
#     data.update!(processed: result)
#   end
# end

# Example: ActiveRecord Extension
# module Processable
#   extend ActiveSupport::Concern
#
#   included do
#     after_save :process_data, if: :should_process?
#   end
#
#   def process_data
#     module = Elide.require('./elide-array-unique.ts')
#     self.processed_value = module.default(self.raw_value)
#     save!
#   end
#
#   def should_process?
#     raw_value_changed?
#   end
# end

# Example: Caching with Redis
# class CachedProcessor
#   def initialize
#     @module = Elide.require('./elide-array-unique.ts')
#     @redis = Redis.new
#   end
#
#   def process(data)
#     cache_key = Digest::SHA256.hexdigest(data.to_s)
#     cached = @redis.get(cache_key)
#     return JSON.parse(cached) if cached
#
#     result = @module.default(data)
#     @redis.setex(cache_key, 3600, result.to_json)
#     result
#   end
# end

# Example: Error Handling Wrapper
# class SafeProcessor
#   def initialize
#     @module = Elide.require('./elide-array-unique.ts')
#     @logger = Logger.new(STDOUT)
#   end
#
#   def process(data)
#     @logger.info "Processing data"
#     result = @module.default(data)
#     @logger.info "Success"
#     { success: true, result: result }
#   rescue StandardError => e
#     @logger.error "Error: #{e.message}"
#     { success: false, error: e.message }
#   end
# end

# Example: Testing with RSpec
# RSpec.describe "Module Integration" do
#   let(:module) { Elide.require('./elide-array-unique.ts') }
#
#   describe '#default' do
#     it 'processes data correctly' do
#       result = module.default(test_data)
#       expect(result).to be_present
#     end
#
#     it 'handles edge cases' do
#       expect { module.default(nil) }.to raise_error(TypeError)
#     end
#   end
# end

puts "\n" + "="*60
puts "Extended Examples Available Above"
puts "="*60
puts "\nThese examples demonstrate:"
puts "  • Rails service objects"
puts "  • Controller integration"
puts "  • Background jobs (ActiveJob)"
puts "  • ActiveRecord extensions"
puts "  • Redis caching"
puts "  • Error handling wrappers"
puts "  • RSpec testing"
puts "\nTotal: 7 comprehensive Rails patterns"

File: conversions/array-unique/elide-array-unique.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-array-unique

This demonstrates calling the TypeScript unique implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One unique library shared across TypeScript and Python
- Consistent deduplication across services
- No need for custom Python set conversion logic
- Guaranteed order preservation
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# unique = require('./elide-array-unique.ts')

print("=== Python Calling TypeScript Array Unique ===\n")

# Example 1: Remove duplicates
# data = [1, 2, 2, 3, 3, 3, 4]
# unique_data = unique.default(data)
# print(unique_data)  # [1, 2, 3, 4]

# Example 2: Data Cleaning
# def clean_dataset(records: list) -> list:
#     """Remove duplicate records"""
#     return unique.default(records)

# Example 3: API Response Deduplication
# def get_unique_tags(articles: list) -> list:
#     """Get unique tags from articles"""
#     all_tags = [tag for article in articles for tag in article['tags']]
#     return unique.default(all_tags)

print("Real-world use case:")
print("- Python data pipeline removes duplicate entries")
print("- Uses same TypeScript implementation as Node.js API")
print("- Guarantees consistent deduplication behavior")
print("- No custom Python unique logic needed")
print()

print("Example: Data Deduplication Pipeline")
print("┌─────────────────────────────────────────────┐")
print("│   Elide Array Unique (TypeScript)         │")
print("│   elide-array-unique.ts                    │")
print("└─────────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │Pipeline│")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same deduplication everywhere!")
print()

print("Problem Solved:")
print("Before: Different unique implementations = inconsistent results")
print("After: One Elide implementation = consistent deduplication")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  unique = require('./elide-array-unique.ts')")
print("  result = unique.default([1, 2, 2, 3])  # [1, 2, 3]")


# ============================================================
# Extended Usage Examples
# ============================================================

# Example: Data Processing Pipeline
# class DataProcessor:
#     def __init__(self):
#         self.module = require('./elide-array-unique.ts')
#     
#     def process_batch(self, data):
#         """Process data using Elide module"""
#         return self.module.default(data)
#     
#     def process_stream(self, stream):
#         """Process streaming data"""
#         results = []
#         for batch in stream:
#             results.append(self.process_batch(batch))
#         return results

# Example: API Integration
# from flask import Flask, request, jsonify
# 
# app = Flask(__name__)
# module = require('./elide-array-unique.ts')
# 
# @app.route('/api/process', methods=['POST'])
# def process_data():
#     data = request.json.get('data')
#     result = module.default(data)
#     return jsonify({'result': result})

# Example: Async Processing
# import asyncio
# 
# async def process_async(data):
#     """Async wrapper for Elide module"""
#     loop = asyncio.get_event_loop()
#     return await loop.run_in_executor(None, module.default, data)
# 
# async def process_multiple(data_list):
#     """Process multiple items concurrently"""
#     tasks = [process_async(data) for data in data_list]
#     return await asyncio.gather(*tasks)

# Example: Caching Layer
# from functools import lru_cache
# 
# @lru_cache(maxsize=1000)
# def process_cached(data_tuple):
#     """Cached processing for frequently used inputs"""
#     return module.default(list(data_tuple))

# Example: Error Handling
# def safe_process(data):
#     """Process with comprehensive error handling"""
#     try:
#         if not data:
#             raise ValueError("Empty data provided")
#         result = module.default(data)
#         return {'success': True, 'result': result}
#     except TypeError as e:
#         return {'success': False, 'error': f'Type error: {e}'}
#     except Exception as e:
#         return {'success': False, 'error': f'Unexpected error: {e}'}

# Example: Logging and Monitoring
# import logging
# import time
# 
# logger = logging.getLogger(__name__)
# 
# def process_with_logging(data):
#     """Process with performance logging"""
#     start = time.time()
#     logger.info(f"Processing data of size: {len(data)}")
#     
#     result = module.default(data)
#     
#     duration = time.time() - start
#     logger.info(f"Processed in {duration:.4f}s")
#     return result

# Example: Testing Utilities
# import unittest
# 
# class TestModule(unittest.TestCase):
#     def setUp(self):
#         self.module = require('./elide-array-unique.ts')
#     
#     def test_basic_functionality(self):
#         """Test basic module functionality"""
#         result = self.module.default(test_data)
#         self.assertIsNotNone(result)
#     
#     def test_edge_cases(self):
#         """Test edge case handling"""
#         self.assertRaises(TypeError, self.module.default, None)

# Example: Configuration Management
# class ConfigurableProcessor:
#     def __init__(self, config):
#         self.config = config
#         self.module = require('./elide-array-unique.ts')
#     
#     def process(self, data):
#         """Process data according to configuration"""
#         if self.config.get('validate'):
#             self.validate_data(data)
#         
#         result = self.module.default(data)
#         
#         if self.config.get('transform'):
#             result = self.transform_result(result)
#         
#         return result
#     
#     def validate_data(self, data):
#         """Validate input data"""
#         pass
#     
#     def transform_result(self, result):
#         """Transform output"""
#         return result

print("\n" + "="*60)
print("Extended Examples Available Above")
print("="*60)
print("\nThese examples demonstrate:")
print("  • Batch processing")
print("  • API integration (Flask)")
print("  • Async/await patterns")
print("  • Caching strategies")
print("  • Error handling")
print("  • Logging and monitoring")
print("  • Unit testing")
print("  • Configuration management")
print("\nTotal: 8 comprehensive patterns")
